larpandoracontent/LArTrackShowerId/SvmPfoCharacterisationAlgorithm.cc

```diff
@@ -35,42 +40,79 @@ bool SvmPfoCharacterisationAlgorithm::IsClearTrack(const Cluster *const pCluster
     if (pCluster->GetNCaloHits() < m_minCaloHitsCut)
         return false;
 
-    SupportVectorMachine::DoubleVector featureVector(LArSvmHelper::CalculateFeatures(m_featureToolVector, this, pCluster));
+    const SupportVectorMachine::DoubleVector featureVector(LArSvmHelper::CalculateFeatures(m_featureToolVector, this, pCluster));
 
-    if (m_ratioVariables)
+    if (m_trainingSetMode)
     {
-        // TODO This assumption is very bad - remove
-        const double straightLineLength(featureVector.at(0));
+        bool isTrueTrack(false);
 
-        if (straightLineLength > std::numeric_limits<double>::epsilon())
+        try
         {
-            for (unsigned int i = 1; i < featureVector.size(); ++i)
-                featureVector[i] /= straightLineLength;
+            const MCParticle *const pMCParticle(MCParticleHelper::GetMainMCParticle(pCluster));
+            isTrueTrack = ((PHOTON != pMCParticle->GetParticleId()) && (E_MINUS != std::abs(pMCParticle->GetParticleId())));
         }
+        catch (const StatusCodeException &) {}
+
+        LArSvmHelper::ProduceTrainingExample(m_trainingOutputFile, isTrueTrack, featureVector);
+        return isTrueTrack;
+    }
+
+    if (!m_enableProbability)
+    {
+        return LArSvmHelper::Classify(m_supportVectorMachine, featureVector);
+    }
+    else
+    {
+        return (LArSvmHelper::CalculateProbability(m_supportVectorMachine, featureVector) > m_minProbabilityCut);
     }
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool SvmPfoCharacterisationAlgorithm::IsClearTrack(const pandora::ParticleFlowObject *const pPfo) const
+{
+
+    if (!LArPfoHelper::IsThreeD(pPfo))
+        return (pPfo->GetParticleId() == MU_MINUS);
```

Comment:
So everything is a track if it doesn't have any 3D hits? Are we sure this is the right thing to do? Maybe redirect to 3x2D (although there might only be e.g. 1x2D or 2x2D available in this case!)?